src/main/scala/com/fulcrumgenomics/umi/CollectDuplexSeqMetrics.scala

```diff
@@ -0,0 +1,540 @@
+    val description = this.description.getOrElse {
+      val samples   = in.readGroups.map(_.getSample).distinct
+      val libraries = in.readGroups.map(_.getLibrary).distinct
+      (samples, libraries) match {
+        case (Seq(sample), Seq(library)) => s"$sample / $library"
+        case _                           => input.getFileName.toString
+      }
+    }
+
+    // Do a bunch of metrics collection
+    collect(iterator)
+    in.close()
+
+    // Write the output files
+    write(description)
+  }
+
+  /** Consumes from the iterator and collects information internal from which to generate metrics. */
+  def collect(iterator: Iterator[SamRecord]): Unit = {
+    val buffered = iterator.bufferBetter
+    val progress = ProgressLogger(logger)
+
+    while (buffered.hasNext) {
+      val group = takeNextGroup(buffered)
+
+      // Assign reads a random number between 0 and 1 inclusive based on their read name
+      group.foreach { rec =>
+        val intHash    = math.abs(this.hasher.hashUnencodedChars(rec.name))
+        val doubleHash = intHash / MaxIntAsDouble
+        rec.transientAttrs(HashKey) = doubleHash
+      }
+
+      // Update the counters
+      this.dsLevels.foreach { fraction =>
+        val downsampledGroup = group.filter(_.transientAttrs[Double](HashKey) <= fraction)
+
+        if (downsampledGroup.nonEmpty) {
+          this.startStopFamilyCounter(fraction).count(downsampledGroup.size)
+          val dsGroups = downsampledGroup.groupBy(r => r[String](MI).takeWhile(_ != '/')).values.toSeq
+          dsGroups.foreach { dsGroup =>
+            // Family sizes first
+            val ssGroups = dsGroup.groupBy(r => r[String](MI)).values.toSeq
+            val counts = ssGroups.map(_.size).sortBy(x => -x)
+            val ab = counts.head
+            val ba = if (counts.length == 2) counts(1) else 0
+            duplexFamilyCounter(fraction).count(Pair(ab, ba))
+
+            // Then the UMIs
+            if (fraction == 1.0) updateUmiMetrics(ssGroups)
+          }
+        }
+      }
+
+      // Record progress
+      group.foreach(progress.record)
+    }
+  }
+
+  /** Updates all the various metrics for the UMIs contained on the given records. */
+  private def updateUmiMetrics(ssGroups: Seq[Seq[SamRecord]]): Unit = {
+    val (ab, ba) = ssGroups match {
+      case Seq(a, b) => (a, b)
+      case Seq(a)    => (a, Seq.empty)
+      case _         => unreachable(s"Found ${ssGroups.size} single strand families in a double strand family!")
+    }
+
+    val umi1s = new ArrayBuffer[String]
+    val umi2s = new ArrayBuffer[String]
+
+    ab.iterator.map(r => r[String](RX).split('-')).foreach { case Array(u1, u2) => umi1s += u1; umi2s += u2 }
+    ba.iterator.map(r => r[String](RX).split('-')).foreach { case Array(u1, u2) => umi1s += u2; umi2s += u1 }
+
+    Seq(umi1s, umi2s).foreach { umis =>
+      val consensus = callConsensus(umis)
+      if (consensus.contains('N')) {
+        val x = 0
+      }
```

Comment:
Oops, I think I had that in there so I could catch and set a breakpoint, I'll pull it back out.